consensus/checks.go

```diff
@@ -166,3 +184,70 @@ func (consensus *Consensus) onPreparedSanityChecks(
 
 	return true
 }
+
+func (consensus *Consensus) viewChangeSanityCheck(msg *msg_pb.Message) bool {
+	senderKey, err := consensus.verifyViewChangeSenderKey(msg)
+	if err != nil {
+		consensus.getLogger().Error().Err(err).Msgf(
+			"[%s] VerifySenderKey Failed",
+			msg.GetType().String(),
+		)
+		return false
+	}
+	if err := verifyMessageSig(senderKey, msg); err != nil {
+		consensus.getLogger().Error().Err(err).Msgf(
+			"[%s] Failed To Verify Sender's Signature",
+			msg.GetType().String(),
+		)
+		return false
+	}
+	return true
+}
+
+func (consensus *Consensus) onViewChangeSanityCheck(recvMsg *FBFTMessage) bool {
+	// TODO: if difference is only one, new leader can still propose the same committed block to avoid another view change
+	// TODO: new leader catchup without ignore view change message
+	if consensus.blockNum > recvMsg.BlockNum {
+		consensus.getLogger().Debug().
+			Uint64("MsgBlockNum", recvMsg.BlockNum).
+			Msg("[onViewChange] Message BlockNum Is Low")
+		return false
+	}
+	if consensus.blockNum < recvMsg.BlockNum {
+		consensus.getLogger().Warn().
+			Uint64("MsgBlockNum", recvMsg.BlockNum).
+			Msg("[onViewChange] New Leader Has Lower Blocknum")
+		return false
+	}
+	if consensus.current.Mode() == ViewChanging &&
+		consensus.current.ViewID() > recvMsg.ViewID {
+		consensus.getLogger().Warn().
+			Uint64("MyViewChangingID", consensus.current.ViewID()).
+			Uint64("MsgViewChangingID", recvMsg.ViewID).
+			Msg("[onViewChange] ViewChanging ID Is Low")
+		return false
+	}
+	if recvMsg.ViewID-consensus.current.ViewID() > MaxViewIDDiff {
```

Comment:
is this the only new condition added?

Reply:
for `onViewChange` sanity check, yes. I also added several other conditions:
* `onNewViewSanityCheck` is new
* moved common checks into `viewChangeSanityCheck`
* added `if consensus.current.Mode() == Normal || consensus.current.Mode() == Syncing {` in `onPreparedSanityChecks`
* added `} else if recvMsg.BlockNum-consensus.blockNum > MaxBlockNumDiff {` in `isRightBlockNumCheck`
* added 
```if consensus.current.Mode() == ViewChanging {
				consensus.getLogger().Debug().Msg(
					"[OnAnnounce] Already in ViewChanging mode, conflicing announce, doing noop",
				)
			}``` in `onAnnounceSanityChecks`
* and some parameterized log messages based on message types.